### PR TITLE
Initial Sender Implementation

### PR DESCRIPTION
**Pull request by jaiveerk**

### CODE REVIEW

File: v2/sender/sender.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"reflect"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
+)
+
+// MessageBody is a type to represent that an input message body can be of any type
+type MessageBody any
+
+// SendHandlerFunction is a function that is used to modify the ServiceBus message before sending
+type SendHandlerFunction func(ctx context.Context, msg *azservicebus.Message)
```
Comment (serbrech) [inline v2/sender/sender.go:15]: `SendHandlerFunc`

File: v2/sender/sender.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+// CustomSender is a wrapper around the ServiceBus sender that allows for users to introduce middleware to modify the ServiceBus message before it's sent
+type CustomSender interface {
+	SendMessage(ctx context.Context, mb MessageBody, options *azservicebus.SendMessageOptions, middleware SendHandlerFunction) error
+}
+
+// Sender is an interface that allows users to substitute their own senders for the default ServiceBus sender when making a CustomSender
```
Comment (serbrech) [inline v2/sender/sender.go:22]: // Sender is satisfied by *azservicebus.Sender

Comment (serbrech) [reply v2/sender/sender.go:22]: maybe call this SBSender to not conflict with goshuttle.Sender?

File: v2/sender/sender.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+// DefaultSender contains a Sender used to send the message to the ServiceBus queue and a Marshaller used to marshal any struct into a ServiceBus message
+type DefaultSender struct {
+	sender     Sender
+	marshaller Marshaller
+}
+
+var _ CustomSender = &DefaultSender{}
+
+// NewDefaultSender takes in a Sender and a Marshaller to create a new object that can send messages to the ServiceBus queue
+func NewDefaultSender(sender Sender, marshaller Marshaller) *DefaultSender {
```
Comment (serbrech) [inline v2/sender/sender.go:36]: NewSender

Comment (serbrech) [reply v2/sender/sender.go:36]: take options, not Marshaller directly

File: v2/sender/sender.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	return &DefaultSender{sender: sender, marshaller: marshaller}
+}
+
+// SendMessage marshals the input struct, runs middleware to modify the returned ServiceBus message, and uses the Sender to send the message to the ServiceBus queue
+func (d *DefaultSender) SendMessage(ctx context.Context, mb MessageBody, options *azservicebus.SendMessageOptions, middleware SendHandlerFunction) error {
```
Comment (serbrech) [inline v2/sender/sender.go:41]: since we apply a middleware func on the SendMessage, let's roll the azserviecus.SendMessageOptions as a SendHandlerFunc to clean up the function signature

Comment (serbrech) [inline v2/sender/sender.go:41]: rename `middleware SendHandlerFunction`
to options SendOption

File: v2/sender/sender.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+		return fmt.Errorf("failed to send message: %s", err)
+	}
+
+	return nil
+}
+
+// SetTypeHandler sets the ServiceBus message's type to the original struct's type
+func SetTypeHandler(mb MessageBody, handler SendHandlerFunction) SendHandlerFunction {
+	return func(ctx context.Context, msg *azservicebus.Message) {
+		var msgType string
```
Comment (serbrech) [inline v2/sender/sender.go:64]: should we implement this as a func on the MessageBody type?

`func (mb MessageBody) GetMessageType() string`

File: v2/sender/marshaller.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
+package sender
```
Comment (serbrech) [inline v2/sender/marshaller.go:1]: can we try to keep this at the root of the package?
if there are things we don't want to expose, we can move them in an `internal` package in a refactoring

**Review by serbrech — CHANGES_REQUESTED** [review]
few comments. direction is about right 👍

File: v2/sender/sender.go (hunk 5 above)
Comment (jaiveerk) [reply v2/sender/sender.go:64]: Unless I'm missing something, we can't make it a method since the MessageBody type is an interface, but I can still pull this out to a `GetMessageType(mb MessageBody) string` function?

File: v2/sender/sender.go (hunk 3 above)
Comment (jaiveerk) [reply v2/sender/sender.go:36]: I'm not sure I understand -
do you mean to take a SendOption (aka a handler/middleware)? If we do that, we don't have a guaranteed step in Send that converts the message body to a SB message, and so we can't feed the message/messagebody into the middleware chain.
Or, do you mean to abstract the Marshaller type into a separate "option" interface that has a method to marshal the messagebody into a sb message?

Comment (jaiveerk) [reply v2/sender/sender.go:36]: Oh, looking at the v1 code I think I understand - if I understand correctly, we should have a SenderOptions field that contains a marshaller? And then construct the sender with that

File: v2/sender/sender.go (hunk 2 above)
Comment (jaiveerk) [reply v2/sender/sender.go:22]: renamed the field to be sbSender as well

File: .github/workflows/v2-build-and-ut.yaml [hunk 7]
```diff
@@ -18,6 +18,11 @@ jobs:
         with:
           go-version: 1.18
 
+      - name: golangci-lint
```
Comment (serbrech) [inline .github/workflows/v2-build-and-ut.yaml:21]: 👍 , but split this out in a separate PR please

File: v2/sender.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+var _ CustomSender = &Sender{}
+
+// NewSender takes in a Sender and a Marshaller to create a new object that can send messages to the ServiceBus queue
+func NewSender(sender SBSender, options SenderOptions) *Sender {
+	return &Sender{sbSender: sender, options: options}
+}
+
+// SendMessage marshals the input struct, runs middleware to modify the returned ServiceBus message, and uses the Sender to send the message to the ServiceBus queue
+func (d *Sender) SendMessage(ctx context.Context, mb MessageBody, options SendOption) error {
```
Comment (serbrech) [inline v2/sender.go:45]: `options ...SendOptions`

File: v2/sender.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	return nil
+}
+
+func (s SenderOptions) Marshaller() Marshaller {
+	return s.marshaller
+}
+
+// SetTypeHandler sets the ServiceBus message's type to the original struct's type
+func SetTypeHandler(mb MessageBody) func(ctx context.Context, msg *azservicebus.Message) error {
```
Comment (serbrech) [inline v2/sender.go:69]: SetMessageType

File: v2/sender.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+// SetMessageIdHandler sets the ServiceBus message's ID to a user-specified value
+func SetMessageIdHandler(messageId *string) func(ctx context.Context, msg *azservicebus.Message) error {
+	return func(ctx context.Context, msg *azservicebus.Message) error {
+		msg.MessageID = messageId
+		return nil
+	}
+}
+
+func GetMessageType(mb MessageBody) string {
```
Comment (serbrech) [inline v2/sender.go:86]: private

File: v2/sender.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+func (s SenderOptions) Marshaller() Marshaller {
+	return s.marshaller
+}
+
+// SetTypeHandler sets the ServiceBus message's type to the original struct's type
+func SetTypeHandler(mb MessageBody) func(ctx context.Context, msg *azservicebus.Message) error {
+	return func(ctx context.Context, msg *azservicebus.Message) error {
+		msgType := GetMessageType(mb)
+		msg.ContentType = &msgType
```
Comment (serbrech) [inline v2/sender.go:72]: the message type should be set on a custom property, not on the ContentType

File: v2/sender.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
+)
+
+// MessageBody is a type to represent that an input message body can be of any type
+type MessageBody any
+
+// CustomSender is a wrapper around the ServiceBus sender that allows for users to introduce middleware to modify the ServiceBus message before it's sent
+type CustomSender interface {
+	SendMessage(ctx context.Context, mb MessageBody, options ...func(ctx context.Context, msg *azservicebus.Message) error) error
```
Comment (serbrech) [inline v2/sender.go:16]: remove context from the option func

File: v2/sender.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// SBSender is satisfied by *azservicebus.Sender
+type SBSender interface {
```
Comment (serbrech) [inline v2/sender.go:20]: MessageSender

Comment (serbrech) [reply v2/sender.go:20]: or `AzServiceBusSender`

File: v2/sender.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+type SenderOptions struct {
+	marshaller Marshaller
+}
+
+var _ CustomSender = &Sender{}
+
+// NewSender takes in a Sender and a Marshaller to create a new object that can send messages to the ServiceBus queue
+func NewSender(sender SBSender, options SenderOptions) *Sender {
```
Comment (serbrech) [inline v2/sender.go:37]: pointer

Comment (serbrech) [reply v2/sender.go:37]: if options nil, provide default value

File: v2/sender.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+	// run user-provided middleware
+	for _, option := range options {
+		err := option(ctx, msg)
+		if err != nil {
+			return fmt.Errorf("failed to run middleware: %s", err)
+		}
+	}
+
+	err = d.sbSender.SendMessage(ctx, msg, &azservicebus.SendMessageOptions{}) // sendMessageOptions currently does nothing
```
Comment (serbrech) [inline v2/sender.go:57]: single line err handling

File: v2/sender.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	// uses a marshaller to marshal the message into a service bus message
+	msg, err := d.options.Marshaller().Marshal(mb)
+	if err != nil {
+		return fmt.Errorf("failed to marshal original struct into ServiceBus message: %s", err)
+	}
+
+	// run user-provided middleware
+	for _, option := range options {
+		err := option(ctx, msg)
+		if err != nil {
```
Comment (serbrech) [inline v2/sender.go:52]: single line err handling

Comment (serbrech) [reply v2/sender.go:52]: fmt.Errorf : use %w for error wrapping, not %s

File: v2/sender.go (hunk 11 above)
Comment (jaiveerk) [reply v2/sender.go:72]: moving this to be inside the Send function like we discussed, removing the option func entirely

File: v2/sender.go [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+	msgTypeField = "type"
+)
+
+// MessageBody is a type to represent that an input message body can be of any type
+type MessageBody any
+
+// CustomSender is a wrapper around the ServiceBus sender that allows for users to introduce middleware to modify the ServiceBus message before it's sent
+type CustomSender interface {
+	SendMessage(ctx context.Context, mb MessageBody, options ...func(msg *azservicebus.Message) error) error
+}
```
Comment (serbrech) [inline v2/sender.go:22]: drop this interface, this can be defined user side if they want to.
we shouldn't provide interfaces

File: v2/sender.go [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+func NewSender(sender AzServiceBusSender, options *SenderOptions) *Sender {
+	if options == nil {
+		options = &SenderOptions{marshaller: &DefaultJSONMarshaller{}}
+	}
+	return &Sender{sbSender: sender, options: options}
+}
+
+// SendMessage marshals the input struct, runs middleware to modify the returned ServiceBus message, and uses the Sender to send the message to the ServiceBus queue
+func (d *Sender) SendMessage(ctx context.Context, mb MessageBody, options ...func(msg *azservicebus.Message) error) error {
+	// uses a marshaller to marshal the message into a service bus message
```
Comment (serbrech) [inline v2/sender.go:51]: unecessary comment, repeats the code. remove it

File: v2/sender.go [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+	msg, err := d.options.Marshaller().Marshal(mb)
+	if err != nil {
+		return fmt.Errorf("failed to marshal original struct into ServiceBus message: %w", err)
+	}
+
+	// set type in application properties
```
Comment (serbrech) [inline v2/sender.go:57]: same. remove comment

File: v2/sender.go [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+	msgType := getMessageType(mb)
+	msg.ApplicationProperties = map[string]interface{}{msgTypeField: msgType}
+
+	// run user-provided msg options
```
Comment (serbrech) [inline v2/sender.go:61]: and same here :)

**Review by serbrech — APPROVED** [review]
LGTM
